integration/connector_test.go

```diff
@@ -0,0 +1,522 @@
+	TenantID string `json:"tenantId"`
+	PolicyID string `json:"policyId"`
+}
+
+type ConnectorSuite struct {
+	suite.Suite
+
+	mqttClient  MQTT.Client
+	dittoClient *ditto.Client
+
+	cfg      *testConfig
+	thingCfg *thingConfig
+
+	thingURL   string
+	featureURL string
+}
+
+const (
+	featureID               = "ConnectorTestFeature"
+	propertyName            = "testProperty"
+	commandName             = "testCommand"
+	responsePayloadTemplate = "responsePayload: %s"
+	https                   = "https"
+	httpsDefaultPort        = "443"
+	httpDefaultPort         = "80"
+)
+
+func (suite *ConnectorSuite) SetupSuite() {
+	cfg := &testConfig{}
+
+	suite.T().Log(getConfigHelp(*cfg))
+
+	if err := initConfigFromEnv(cfg); err != nil {
+		suite.T().Fatal(err)
+	}
+
+	suite.T().Logf("test config: %+v", *cfg)
+
+	opts := MQTT.NewClientOptions().
+		AddBroker(cfg.Broker).
+		SetClientID(uuid.New().String()).
+		SetKeepAlive(30 * time.Second).
+		SetCleanSession(true).
+		SetAutoReconnect(true)
+
+	mqttClient := MQTT.NewClient(opts)
+
+	if token := mqttClient.Connect(); token.Wait() && token.Error() != nil {
+		require.NoError(suite.T(), token.Error(), "connect to MQTT broker")
+	}
+
+	thingCfg, err := getThingConfig(mqttClient)
+	if err != nil {
+		mqttClient.Disconnect(uint(cfg.MqttQuiesceMs))
+		require.NoError(suite.T(), err, "get thing config")
+	}
+
+	suite.T().Logf("thing config: %+v", *thingCfg)
+
+	dittoClient, err := ditto.NewClientMQTT(mqttClient, ditto.NewConfiguration())
+	if err == nil {
+		err = dittoClient.Connect()
+	}
+
+	if err != nil {
+		mqttClient.Disconnect(uint(cfg.MqttQuiesceMs))
+		require.NoError(suite.T(), err, "initialize ditto client")
+	}
+
+	feature := &model.Feature{}
+	feature.WithProperty(propertyName, "testValue")
+
+	cmd := things.NewCommand(model.NewNamespacedIDFrom(thingCfg.DeviceID)).Twin().Feature(featureID).
+		Modify(feature)
+	msg := cmd.Envelope(protocol.WithResponseRequired(false))
+
+	err = dittoClient.Send(msg)
+	require.NoError(suite.T(), err, "create test feature")
+
+	suite.dittoClient = dittoClient
+	suite.mqttClient = mqttClient
+	suite.cfg = cfg
+	suite.thingCfg = thingCfg
+
+	suite.thingURL = fmt.Sprintf("%s/api/2/things/%s", strings.TrimSuffix(cfg.DigitalTwinAPIAddress, "/"), thingCfg.DeviceID)
+	suite.featureURL = fmt.Sprintf("%s/features/%s", suite.thingURL, featureID)
+}
+
+func (suite *ConnectorSuite) TearDownSuite() {
+	if _, err := suite.doRequest(http.MethodDelete, suite.featureURL); err != nil {
+		suite.T().Logf("error while deleting test feature: %v", err)
+	}
+
+	suite.dittoClient.Disconnect()
+	suite.mqttClient.Disconnect(uint(suite.cfg.MqttQuiesceMs))
+}
+
+func TestConnectorSuite(t *testing.T) {
+	suite.Run(t, new(ConnectorSuite))
+}
+
+func (suite *ConnectorSuite) TestConnectionStatus() {
+	type connectionStatus struct {
+		ReadySince time.Time `json:"readySince"`
+		ReadyUntil time.Time `json:"readyUntil"`
+	}
+
+	timeout := time.Duration(suite.cfg.StatusTimeoutMs * int(time.Millisecond))
```

Comment:
As it turns out, these are prone to overflow on 32-bit systems: `time.Millisecond` is in nanoseconds, so some of these can very easily pass the limits of an int. (e.g. the MQTT event timeout is 30 billion nanoseconds, and the status timeout is 10 billion). This results in some very nice errors:

```
2022-11-09 20:45:20.517717   OUT =>             Error:          Received unexpected error:
2022-11-09 20:45:20.518031   OUT =>                             WS response not received in -64.771072ms, last error: <nil>
```

Patch to fix:

```
diff --git a/integration/connector_test.go b/integration/connector_test.go
index 7754d0e..21e15a6 100644
--- a/integration/connector_test.go
+++ b/integration/connector_test.go
@@ -164,11 +164,11 @@ func (suite *ConnectorSuite) TestConnectionStatus() {
 		ReadyUntil time.Time `json:"readyUntil"`
 	}
 
-	timeout := time.Duration(suite.cfg.StatusTimeoutMs * int(time.Millisecond))
+	timeout := time.Duration(suite.cfg.StatusTimeoutMs) * time.Millisecond
 	threshold := time.Now().Add(timeout)
 
 	firstTime := true
-	sleepDuration := time.Duration(suite.cfg.TimeSleepMs * int(time.Millisecond))
+	sleepDuration := time.Duration(suite.cfg.TimeSleepMs) * time.Millisecond
 	for {
 		if !firstTime {
 			time.Sleep(sleepDuration)
@@ -283,7 +283,7 @@ func (suite *ConnectorSuite) TestCommand() {
 	err = websocket.JSON.Send(ws, cmdMsgEnvelope)
 	require.NoError(suite.T(), err, "unable to send command to the backend via websocket")
 
-	commandTimeout := time.Duration(suite.cfg.EventTimeoutMs * int(time.Millisecond))
+	commandTimeout := time.Duration(suite.cfg.EventTimeoutMs) * time.Millisecond
 	timeoutCh := beginWait(commandTimeout, commandResponseCh, func() {})
 	require.NoError(suite.T(), <-timeoutCh, "command should be received and response should be sent")
 	require.NoError(suite.T(), <-respCh, "command response should be received")
@@ -360,7 +360,7 @@ func (suite *ConnectorSuite) sendDittoEvent(topic string, message interface{}) e
 		return err
 	}
 	token := suite.mqttClient.Publish(topic, 1, false, payload)
-	timeout := time.Duration(suite.cfg.MqttAcknowledgeTimeoutMs * int(time.Millisecond))
+	timeout := time.Duration(suite.cfg.MqttAcknowledgeTimeoutMs) * time.Millisecond
 	if !token.WaitTimeout(timeout) {
 		return ditto.ErrAcknowledgeTimeout
 	}
@@ -384,7 +384,7 @@ func beginWait(timeout time.Duration, resultCh chan error, closer func()) chan e
 }
 
 func (suite *ConnectorSuite) beginWSWait(ws *websocket.Conn, check func(payload []byte) error) chan error {
-	timeout := time.Duration(suite.cfg.EventTimeoutMs * int(time.Millisecond))
+	timeout := time.Duration(suite.cfg.EventTimeoutMs) * time.Millisecond
 	resultCh := make(chan error)
 
 	go func() {

```